data/lua/wml-tags.lua

```diff
@@ -423,12 +442,50 @@ wml_actions["while"] = function( cfg )
 	for i = 1, 65536 do
 		if wesnoth.eval_conditional( cfg ) then
 			for do_child in helper.child_range( cfg, "do" ) do
-				handle_event_commands( do_child )
+				local action = handle_event_commands(do_child, "loop")
+				if action == "break" or action == "return" then return end
 			end
 		else return end
 	end
 end
 
+wesnoth.wml_actions["for"] = function(cfg)
+	local first, last, step
+	if cfg.array then
+		first = 0
+		last = wesnoth.get_variable(cfg.array .. ".length") - 1
+		step = 1
+		if cfg.reverse == "yes" then
+			first, last = last, first
+			step = -1
+		end
+	else
+		first = cfg.start or 0
+		last = cfg["end"] or first
+		step = cfg.step or ((last - first) / math.abs(last - first))
+	end
+	local i_var = cfg.variable or "i"
+	local save_i = start_var_scope(i_var)
+	wesnoth.set_variable(i_var, first)
+	while wesnoth.get_variable(i_var) <= last do
+		for do_child in helper.child_range( cfg, "do" ) do
+			local action = handle_event_commands(do_child, "loop")
+			if action == "break" or action == "return" then return end
+		end
+		wesnoth.set_variable(i_var, wesnoth.get_variable(i_var) + 1)
+	end
+	end_var_scope(i_var, save_i)
+end
+
+wml_actions["repeat"] = function(cfg)
+	local times = cfg.times or 65536
```

Comment:
Is there precedent for 65536? Anyone who tries to use this as an infinite loop will be very surprised after a long enough time.